earthfile2llb/interpreter.go

```diff
@@ -1027,31 +1027,47 @@ func (i *Interpreter) handleEnv(ctx context.Context, cmd spec.Command) error {
 	return nil
 }
 
-func (i *Interpreter) handleArg(ctx context.Context, cmd spec.Command) error {
-	if i.pushOnlyAllowed {
-		return i.pushOnlyErr(cmd.SourceLocation)
-	}
+var errInvalidSyntax = errors.New("invalid syntax")
+var errRequiredArgHasDefault = errors.New("required ARG cannot have a default value")
+
+// parseArgArgs parses the ARG command's arguments
+// and returns the argOpts, key, value (or nil if missing), or error
+func parseArgArgs(ctx context.Context, cmd spec.Command) (argOpts, string, *string, error) {
 	opts := argOpts{}
 	args, err := flagutil.ParseArgs("ARG", &opts, getArgsCopy(cmd))
 	if err != nil {
-		return i.wrapError(err, cmd.SourceLocation, "invalid ARG arguments %v", cmd.Args)
+		return argOpts{}, "", nil, err
 	}
-	var key, value string
 	switch len(args) {
 	case 3:
 		if args[1] != "=" {
-			return i.errorf(cmd.SourceLocation, "invalid syntax")
+			return argOpts{}, "", nil, errInvalidSyntax
 		}
-		value = i.expandArgs(args[2], true)
-		fallthrough
-	case 1:
-		if opts.Required && len(value) != 0 {
-			return i.errorf(cmd.SourceLocation, "required ARG cannot have a default value")
+		if opts.Required {
+			return argOpts{}, "", nil, errRequiredArgHasDefault
 		}
-		key = args[0] // Note: Not expanding args for key.
+		return opts, args[0], &args[2], nil
+	case 1:
+		return opts, args[0], nil, nil
 	default:
-		return i.errorf(cmd.SourceLocation, "invalid syntax")
+		return argOpts{}, "", nil, errInvalidSyntax
 	}
+}
+
+func (i *Interpreter) handleArg(ctx context.Context, cmd spec.Command) error {
+	if i.pushOnlyAllowed {
+		return i.pushOnlyErr(cmd.SourceLocation)
+	}
+	opts, key, valueOrNil, err := parseArgArgs(ctx, cmd)
+	if err != nil {
+		return i.wrapError(err, cmd.SourceLocation, "invalid ARG arguments %v", cmd.Args)
+	}
+
+	var value string
+	if valueOrNil != nil {
```

Comment:
If you returned an empty-string from `parseArgArgs` instead would it be sufficient to just check `value != ""` here instead?

I think the pointer adds a _slight_ additional cost plus potential greater chance for panic in code if someone changes the code and mistakenly forgets to guard against nil somewhere

Reply:
my idea was to differentiate between `ARG key` and `ARG key = ""`. I figured a nil pointer kept the number of returned arguments smaller rather than introducing a `valueExists` bool. I could expand it to 5 returned values if you think that's more clear.

Reply:
I think pointer makes more sense knowing now that you need to differentiate between empty string and lack of string (and 5 return values would be a bit much)